src/server/cluster/cluster_config.h

```diff
@@ -3,36 +3,83 @@
 //
 
 #pragma once
-#include <absl/container/flat_hash_set.h>
 
+#include <absl/base/thread_annotations.h>
+
+#include <array>
+#include <optional>
 #include <string_view>
-#include <tuple>
+#include <vector>
 
 #include "src/core/fibers.h"
 
 namespace dfly {
 
-typedef uint16_t SlotId;
+using SlotId = uint16_t;
 
 class ClusterConfig {
  public:
-  ClusterConfig();
+  struct Node {
+    std::string id;
+    std::string ip;
+    uint16_t port = 0;
+  };
+
+  struct SlotRange {
+    SlotId start = 0;
+    SlotId end = 0;
+  };
+
+  struct ClusterShard {
+    std::vector<SlotRange> slot_ranges;
+    Node master;
+    std::vector<Node> replicas;
+  };
+
+  explicit ClusterConfig(std::string_view my_id);
+
   static SlotId KeySlot(std::string_view key);
+
   static bool IsClusterEnabled() {
     return cluster_enabled;
   }
+
   // If the key contains the {...} pattern, return only the part between { and }
   static std::string_view KeyTag(std::string_view key);
 
   // If key is in my slots ownership return true
-  bool IsMySlot(SlotId id);
+  bool IsMySlot(SlotId id) const;
+
+  // Returns nodes that own `id`. Result will always have the first element set to be the master,
+  // and the following 0 or more elements are the replicas.
+  std::vector<Node> GetNodesForSlot(SlotId id) const;
+
+  // Returns the master configured for `id`. Returns a default-initialized `Node` if `SetConfig()`
+  // was never completed successfully.
+  Node GetMasterNodeForSlot(SlotId id) const;
+
+  // Returns true if `new_config` is valid and internal state was changed. Returns false and changes
+  // nothing otherwise.
+  bool SetConfig(const std::vector<ClusterShard>& new_config);
 
  private:
-  void AddSlots();
+  struct SlotOwner {
```

Comment:
In this implementation we are using *16384 space to save the data ,instead saving it only once as the config format, and we will need to have a logic to convert from the slot array to the config format again to be able to reply to the CLISTER SHARDS command.

If you would create a member :
std::vector<ClusterShard> config_; 
Which you will set on Set config command and this will be used to reply CLUSTER SHARDS
The SlotOwner struct can be 
{
string_view ip;
string_view port;
bool owned_by_me;
}

In this implementation if we have one master and one replica, the data will be saved in the config_ member, the slot array will point to the config data.
In the current implementation we are saving the master and the replica data *16384 times

Reply:
I guess that it's a matter of what we're optimizing for. The current optimization is aimed for fast "get owners for slot" questions, such as `MOVED` replies.
If we're looking to save memory, we could consider using `shared_ptr` as the value of the internal map. It has the downside of using atomics, but I guess it's not terrible since it's only going to take effect during `SetConfig()` (I think).
But it's true that the current implementation requires more logic to build the reply of `CLUSTER SHARDS` and `CLUSTER SLOTS`. Not much more logic, essentially building a dictionary by iterating all entries (that'll actually be simpler to implement if we moved to `shared_ptr` as we'll be able to compare `Node`s). But still, we could avoid compressing ranges.
Do you suggest having 2 members (`slots_` and `config_`, where `slots_` will have 16k entries and will point into `config_`)? Because that will keep the 16k times data duplication, but without it I don't think we can guarantee `O(1)` computational time for `MOVED` replies.

Reply:
Yes we should optimise for "get owners for slot" as it is in the flow of client database queries.
But why does my suggestion guarantee O(1) computational time?

Reply:
As discussed offline, I kept 2 members and updated the code according to your suggestion. PTAL.